[quant][pt2e] Actually support transitive sharing for SharedQuantizationSpec

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #111172

Summary:
Previously we actually did not really support this, this PR added the support.

Next
* clean up insert observer logic
* add allow_transitive_sharing boolean flag to allow people to turn this op for certain edges

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_shared_qspec_transitivity

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D50250789](https://our.internmc.facebook.com/intern/diff/D50250789)

jerryzh168 committed:
[quant][pt2e] Actually support transitive sharing for SharedQuantizationSpec

Summary:
Previously we actually did not really support this, this PR added the support.
WIP: putting up the PR for testing first, need more docs

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_shared_qspec_transitivity

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50250789).

jerryzh168 committed:
Update on "[quant][pt2e] Actually support transitive sharing for SharedQuantizationSpec"


Summary:
Previously we actually did not really support this, this PR added the support.
WIP: putting up the PR for testing first, need more docs

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_shared_qspec_transitivity

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D50250789](https://our.internmc.facebook.com/intern/diff/D50250789)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Actually support transitive sharing for SharedQuantizationSpec"


Summary:
Previously we actually did not really support this, this PR added the support.
WIP: putting up the PR for testing first, need more docs

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_shared_qspec_transitivity

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D50250789](https://our.internmc.facebook.com/intern/diff/D50250789)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Actually support transitive sharing for SharedQuantizationSpec"


Summary:
Previously we actually did not really support this, this PR added the support.

Next
* clean up insert observer logic
* add allow_transitive_sharing boolean flag to allow people to turn this op for certain edges

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_shared_qspec_transitivity

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D50250789](https://our.internmc.facebook.com/intern/diff/D50250789)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Actually support transitive sharing for SharedQuantizationSpec"


Summary:
Previously we actually did not really support this, this PR added the support.

Next
* clean up insert observer logic
* add allow_transitive_sharing boolean flag to allow people to turn this op for certain edges

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_shared_qspec_transitivity

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D50250789](https://our.internmc.facebook.com/intern/diff/D50250789)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Actually support transitive sharing for SharedQuantizationSpec"


Summary:
Previously we actually did not really support this, this PR added the support.

Next
* clean up insert observer logic
* add allow_transitive_sharing boolean flag to allow people to turn this op for certain edges

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_shared_qspec_transitivity

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D50250789](https://our.internmc.facebook.com/intern/diff/D50250789)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Actually support transitive sharing for SharedQuantizationSpec"


Summary:
Previously we actually did not really support this, this PR added the support.

Next
* clean up insert observer logic
* add allow_transitive_sharing boolean flag to allow people to turn this op for certain edges

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_shared_qspec_transitivity

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D50250789](https://our.internmc.facebook.com/intern/diff/D50250789)

[ghstack-poisoned]

kimishpatel commented on torch/ao/quantization/pt2e/prepare.py:
```diff
@@ -5,28 +5,223 @@
+    return edge_or_node_to_qspec
+
+def _get_edge_or_node_to_group_id(edge_or_node_to_qspec: Dict[EdgeOrNode, QuantizationSpecBase]) -> Dict[EdgeOrNode, int]:
+    """Map from edge/node to the group ID, generated from quantization annotations,
```
I think it will be also good to document an example where we actually need to enable transitive sharing. And really it is to solve when there is one group with shared spec and another group with shared spec and there an edge between the two. Only way to have an edge between two groups is when there is a node `x` in one group and an edge `(x, y)` in another group. This is basically how x->y is observed was output x is observed.

jerryzh168 replied:
OK, I'll change this to the cat example

kimishpatel commented on torch/ao/quantization/pt2e/prepare.py:
```diff
@@ -5,28 +5,223 @@
+    for edge_or_node, qspec in edge_or_node_to_qspec.items():
+        if isinstance(edge_or_node, torch.fx.Node):
+            output_node = edge_or_node
+            _update_shared_with(output_node, qspec, shared_with_map)
```
Also I thought we were updating transitive sharing as an option in shared quant spec and not make it default. This was to counter the example of add we discussed.

jerryzh168 replied:
yeah, I'm planning to add this in a follow up PR

kimishpatel commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -811,6 +811,152 @@ def validate(self, model: torch.fx.GraphModule) -> None:
+        """This tests the transitivity of SharedQuantizationSpec, that is
+        if A is shared with B, B is shared with C, then C should be shared with A as well
+        """
+        class BackendAQuantizer(Quantizer):
```
we should factor this quantizer out in common util. There is a lot of code duplication here

jerryzh168 replied:
OK, I'll add a TODO

kimishpatel commented on torch/ao/quantization/pt2e/prepare.py:
```diff
@@ -5,28 +5,223 @@
+    root_a = _find_root(a, shared_with_map)
+    root_b = _find_root(b, shared_with_map)
+    # union the two trees
+    shared_with_map[root_b] = root_a
```
can you rename the args a and b to `parent` and `child` because thats what you are doing here.

kimishpatel commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -811,6 +811,152 @@ def validate(self, model: torch.fx.GraphModule) -> None:
+                        input_qspec_map[first_input_node] = act_qspec
+                        share_qparams_with_input_act0_qspec = SharedQuantizationSpec((first_input_node, cat_node))
+                        for input_node in input_nodes[1:]:
+                            input_qspec_map[input_node] = share_qparams_with_input_act0_qspec
```
can we also test when the first input uses shared spec and second input has its own spec?

jerryzh168 replied:
OK sure

kimishpatel commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -811,6 +811,152 @@ def validate(self, model: torch.fx.GraphModule) -> None:
+        m(*example_inputs)
+        m = convert_pt2e(m, fold_quantize=True)
+
+        node_occurrence = {
+            # two for input of the first conv, one for output for the first conv
+            ns.call_function(
+                torch.ops.quantized_decomposed.quantize_per_tensor.default
+            ): 7,
+            ns.call_function(
+                torch.ops.quantized_decomposed.dequantize_per_tensor.default
+            ): 9,
+        }
+        node_list = [
+            ns.call_function(
+                torch.ops.quantized_decomposed.dequantize_per_tensor.default
+            ),
+            ns.call_function(
+                torch.ops.quantized_decomposed.dequantize_per_tensor.default
+            ),
+            ns.call_function(torch.ops.aten.cat.default),
+            ns.call_function(
+                torch.ops.quantized_decomposed.quantize_per_tensor.default
+            ),
+            ns.call_function(
+                torch.ops.quantized_decomposed.dequantize_per_tensor.default
+            ),
+            ns.call_function(torch.ops.aten.cat.default),
+            ns.call_function(
+                torch.ops.quantized_decomposed.quantize_per_tensor.default
+            ),
+        ]
+        self.checkGraphModuleNodes(
+            m, expected_node_list=node_list, expected_node_occurrence=node_occurrence
+        )
```
Do we need to test this?

jerryzh168 replied:
these are just sanity checks, just to make sure q/dq ops are still correct, do you think I should remove?

kimishpatel replied:
Doesnt seem relevant to this test, but sure.

kimishpatel reviewed (changes requested):
looks good overall. left some nits.

jerryzh168 commented on torch/ao/quantization/pt2e/prepare.py:
```diff
@@ -5,28 +5,223 @@
+        return edge_or_node
+    root = _find_root(parent, shared_with_map)
+    # path compression
+    shared_with_map[edge_or_node] = root
```
yeah this is a common optimization in union find, just to make it easier to find the root for the current node

jerryzh168 committed:
Update on "[quant][pt2e] Actually support transitive sharing for SharedQuantizationSpec"


Summary:
Previously we actually did not really support this, this PR added the support.

Next
* clean up insert observer logic
* add allow_transitive_sharing boolean flag to allow people to turn this op for certain edges

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_shared_qspec_transitivity

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D50250789](https://our.internmc.facebook.com/intern/diff/D50250789)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][pt2e] Actually support transitive sharing for SharedQuantizationSpec"


Summary:
Previously we actually did not really support this, this PR added the support.

Next
* clean up insert observer logic
* add allow_transitive_sharing boolean flag to allow people to turn this op for certain edges

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_shared_qspec_transitivity

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D50250789](https://our.internmc.facebook.com/intern/diff/D50250789)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D50250789).

jerryzh168 commented:
@pytorchbot merge
